pkg/resources/network_rule.go

```diff
@@ -0,0 +1,219 @@
+	},
+	"mode": {
+		Type:         schema.TypeString,
+		Required:     true,
+		ForceNew:     true,
+		Description:  "Specifies what is restricted by the network rule. Valid values are INGRESS, INTERNAL_STAGE and EGRESS; see https://docs.snowflake.com/en/sql-reference/sql/create-network-rule#required-parameters for details.",
+		ValidateFunc: validation.StringInSlice([]string{"INGRESS", "INTERNAL_STAGE", "EGRESS"}, false),
+	},
+	"comment": {
+		Type:        schema.TypeString,
+		Optional:    true,
+		Description: "Specifies a comment for the network rule.",
+	},
+	"qualified_name": {
+		Type:        schema.TypeString,
+		Computed:    true,
+		Description: "Qualified name of the network rule.",
+	},
+}
+
+// NetworkRule returns a pointer to the resource representing a network rule.
+func NetworkRule() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: CreateContextNetworkRule,
+		ReadContext:   ReadContextNetworkRule,
+		UpdateContext: UpdateContextNetworkRule,
+		DeleteContext: DeleteContextNetworkRule,
+
+		Schema: networkRuleSchema,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+	}
+}
+
+func CreateContextNetworkRule(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	name := d.Get("name").(string)
+	databaseName := d.Get("database").(string)
+	schemaName := d.Get("schema").(string)
+	id := sdk.NewSchemaObjectIdentifier(databaseName, schemaName, name)
+
+	ruleType := sdk.NetworkRuleType(d.Get("type").(string))
+	ruleMode := sdk.NetworkRuleMode(d.Get("mode").(string))
+
+	valueList := expandStringList(d.Get("value_list").(*schema.Set).List())
+	networkRuleValues := make([]sdk.NetworkRuleValue, len(valueList))
+	for i, v := range valueList {
+		networkRuleValues[i] = sdk.NetworkRuleValue{Value: v}
+	}
+
+	req := sdk.NewCreateNetworkRuleRequest(
+		id,
+		ruleType,
+		networkRuleValues,
+		ruleMode,
+	)
+
+	// Set optionals
+	if v, ok := d.GetOk("comment"); ok {
+		req = req.WithComment(sdk.String(v.(string)))
+	}
+
+	client := meta.(*provider.Context).Client
+	if err := client.NetworkRules.Create(ctx, req); err != nil {
+		return diag.FromErr(err)
+	}
+	d.SetId(helpers.EncodeSnowflakeID(id))
+
+	return ReadContextNetworkRule(ctx, d, meta)
+}
+
+func ReadContextNetworkRule(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	diags := diag.Diagnostics{}
+	client := meta.(*provider.Context).Client
+	id := helpers.DecodeSnowflakeID(d.Id()).(sdk.SchemaObjectIdentifier)
+
+	networkRule, err := client.NetworkRules.ShowByID(ctx, id)
+	if networkRule == nil || err != nil {
+		// If not found, mark resource to be removed from state file during apply or refresh
+		log.Printf("[DEBUG] network rule (%s) not found", d.Id())
+		d.SetId("")
```

Comment:
Done

Reply:
Thank You, the only thing I would change is the error type (that's on me, because the example said `sdk.ErrObjectNotExistOrAuthorized`), because in this case we should compare against `sdk.ErrObjectNotFound` instead.

Reply:
Sure, done.

However, I have questions: What is the exact difference between the two error types and most of all, how to recognize which to use?

Reply:
`ErrObjectNotExistOrAuthorized` is a Snowflake error, so this one could be returned by any operation, and `ErrObjectNotFound` is an SDK error returned from ShowByID operations. It indicates that we called SHOW and filtered its output and didn't find the expected object (you can see ShowByID implementation and that it returns an error inside collections.FindOne function).

Reply:
I understand. Thanks for the explanation :)